Add support and testing matrix to Python standards (#200)
cfpb/development: standards/python.md

We use [isort](https://github.com/timothycrosley/isort) to lint imports to comply with both style guidelines, and provide a [sample isort configuration](../.isort.cfg). This configuration attempts to enforce the following conventions:

- Multiple-line `from` imports become grouped within parenthesis with one imported name per-line and a trailing comma for the last imported name.
- Django imports are included as their own section between the standard library and third-party imports.
- For Wagtail-specific projects like [cfgov-refresh](https://github.com/cfpb/cfgov-refresh), Wagtail imports are included as their own section between Django and third-party imports.

The isort configuration can go into a `tox.ini` file under an `[isort]` header.

## Dependency support/testing matrix

We have two general types of Python projects:

- Applications that have a running instance, e.g., [cfgov-refresh](https://github.com/cfpb/cfgov-refresh)
- Libraries that provide some functionality independent of applications

> We use a the following rubric to determine which versions of Python and major dependencies should be supported. Assuming the project has Django and Wagtail as its major dependencies, our recommendation would be this:

Reviewer: Minor corrections to the text I suggested earlier (removing a typo, and minor wording change for style)

```suggestion
We use a rubric like the one below to determine which versions of Python and major dependencies should be supported. Assuming the project has Django and Wagtail as its major dependencies, our recommendation would be this:
```

|| Library | Application |
|--|--|--|
|Oldest Python 3 supported by Django LTS | ✔️| |
|Current production Python |  | ✔️ |
|Latest Python | ✔️ | ✔️ |
|Latest Django LTS version | ✔️ | ✔️ |
|Latest Django version |  ✔️ | ✔️ |
|Latest Wagtail LTS version | ✔️ | ✔️ |
|Latest Wagtail version |  ✔️ | ✔️ |

Our applications should support the version of Python 3 we currently run in production and the latest version. For example, [cfgov-refresh](https://github.com/cfpb/cfgov-refresh) should support version 3.6 and 3.8 at time of writing.

Our libraries should support the oldest version of Python 3 that the library's major dependencies supports, and the latest version of Python. For example, [Django 2.2 supports Python 3.5 and up](https://docs.djangoproject.com/en/2.2/releases/2.2/#python-compatibility), so our Django and Wagtail libraries that support Django 2.2 should do the same.

> "Major dependencies" are the frameworks upon which our applications and libraries are built. Examples are Django and Wagtail. Both our and our libraries should support the LTS (long-term support) version of that dependency we currently run in production, and the latest version (LTS or not) of that dependency.

Reviewer: ```suggestion
"Major dependencies" are the frameworks upon which our applications and libraries are built. Examples are Django and Wagtail. Both our applications and our libraries should support the LTS (long-term support) version of that dependency we currently run in production, and the latest version (LTS or not) of that dependency.
```

We use [isort](https://github.com/timothycrosley/isort) to lint imports to comply with both style guidelines, and provide a [sample isort configuration](../.isort.cfg). This configuration attempts to enforce the following conventions:

- Multiple-line `from` imports become grouped within parenthesis with one imported name per-line and a trailing comma for the last imported name.
- Django imports are included as their own section between the standard library and third-party imports.
- For Wagtail-specific projects like [cfgov-refresh](https://github.com/cfpb/cfgov-refresh), Wagtail imports are included as their own section between Django and third-party imports.

The isort configuration can go into a `tox.ini` file under an `[isort]` header.

## Dependency support/testing matrix

We have two general types of Python projects:

> - Applications that have a running instance, i.e. [cfgov-refresh](https://github.com/cfpb/cfgov-refresh)

Reviewer 2: ```suggestion
- Applications that have a running instance, e.g., [cfgov-refresh](https://github.com/cfpb/cfgov-refresh)
```

- Libraries that provide some functionality independent of applications

When deciding what versions of Python and major dependencies they should support and be tested against, we use the following rubric:

|| Library | Application |
|--|--|--|

> |Oldest Python 3 supported by major dependencies | ✔️| |

Reviewer: Oldest Python 3 supported by which version of our major dependencies? And if our major dependencies don't support the same oldest version, which do we choose?

Author: If Django supports 3.5 but Wagtail drops 3.5 in favor of 3.6, I think we'd have to have 3.6 as the oldest supported version for a library that depends on Wagtail. 

I'm not sure I know how to write that as a sentence... "in case of a mis-match between dependencies, choose the newest oldest version"? That doesn't work.

Reviewer: What you have in the current draft works for me. If we wanted to be very specific, we could say "Oldest Python 3 supported by the latest LTS of both Django and Wagtail", but I think it's clear enough as is.

> |Current production Python | ✔️ | ✔️ |

Reviewer: "Currently running in production" really only applies to applications, not libraries, right? A library could be in production in many different places on different versions.
```suggestion
|Current production Python |  | ✔️ |
```

Author: I *think* I'm inclined to go with @schbetsy's suggestion. It highlights the difference between expectations of a library and an application, which is the main point of the table?

|Latest Python | ✔️ | ✔️ |

> |Current LTS version of major dependency | ✔️ | ✔️ |

Reviewer: I understand you want this guide to be library-agnostic so other people can use it, but I find the phrase "major dependency" a bit too abstract to make this chart usable. I don't know if this is better, but maybe consider something like this?:

> We use a the following rubric to determine which versions of Python and major dependencies should be supported. Assuming the project has Django and Wagtail as its major dependencies, our recommendation would be this:

|| Library | Application |
|--|--|--|
|Oldest Python 3 supported by Django LTS | ✔️| |
|Current production Python |  | ✔️ |
|Latest Python | ✔️ | ✔️ |
|Latest Django LTS version | ✔️ | ✔️ |
|Latest Django version |  ✔️ | ✔️ |
|Latest Wagtail LTS version | ✔️ | ✔️ |
|Latest Wagtail version |  ✔️ | ✔️ |

Author: Changed per suggestion!

|Latest version of major dependency |  ✔️ | ✔️ |

> Our applications should support the version of Python 3 we currently run in production and the latest version. For example, [cfgov-refresh](https://github.com/cfpb/cfgov-refresh) should support version 3.6 and 3.8.

Reviewer: Consider adding "at time of writing", or something to make it so this sentence won't immediately go out of date.

Author: Done!

When dealing with support for multiple versions of Python or major dependencies, we prefer to [use feature detection instead of version detection](https://docs.python.org/3/howto/pyporting.html#use-feature-detection-instead-of-version-detection) for incompatibilities/breaking changes that may exist between versions.

### Specifying dependencies in `requirements.txt` or `setup.py`

The difference between pinned versions of dependencies in `requirements.txt`  and dependencies specified in `setup.py` is frequently confusing and often misunderstood. The difference is between *concrete* and *abstract* dependencies, which is most easily understood in terms of applications (intended to be deployed and run), which specify *concrete* dependencies, and libraries (reusable code that may be included in applications), which specify *abstract* dependencies.

[Donald Stufft has an excelent article with a more in-depth discussion of this difference](https://caremad.io/posts/2013/07/setup-vs-requirement/). For our purposes, we use both requirements files and requirements specified in `setup.py` depending on the circumstance of the dependency.

> Our applications should use requirements files with pinned versions

Reviewer: Suggestion: Combine lines 75 and 77. ("Our applications should use..." and "For example, cfgov-refresh...")

Author: Done!

> Our applications should use requirements files with pinned versions

Reviewer 2: ```suggestion
Our applications should use requirements files with pinned versions.
```

Author: Fixed per @schbetsy's suggestion above.

> For example, [cfgov-refresh has several requirements](https://github.com/cfpb/cfgov-refresh/tree/master/requirements) files that pin all relevant versions for deployment, testing, local execution, and other specific needs. Some of these requirements files dependend on each other ([`-r libraries.txt`](https://github.com/cfpb/cfgov-refresh/blob/master/requirements/base.txt#L3)).

Reviewer 2: ```suggestion
For example, [cfgov-refresh has several requirements files](https://github.com/cfpb/cfgov-refresh/tree/master/requirements) that pin all relevant versions for deployment, testing, local execution, and other specific needs. Some of these requirements files inherit from each other (e.g., [`base.txt` uses the `-r libraries.txt` directive](https://github.com/cfpb/cfgov-refresh/blob/master/requirements/base.txt#L3) to include the requirements from that file).
```

> Our libraries should specify dependencies with supported version ranges ([and those ranges should be tested with tox](../tox.ini)) in `setup.py` using [the appropriate setuptools `_requires` keyword](https://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-dependencies).

Reviewer: Suggestion: Combine lines 79 and 81 ("Our libraries should specify..." and "For example, wagtail sharing...")

Author: Done!

> 

Reviewer 2: Did you mean to delete this line entirely?

Author: No, I don't think I did, but re-reading it in context, I'm not sure that's accurate, and I'm not sure what to say... satellite apps should probably be treated the same as cfgov-refresh, under this set of conditions, no?

Reviewer 2: Yeah, I was somewhat surprised that they would be considered libraries in the first place. I would agree with treating them like cfgov-refresh.

Reviewer 2: And I think it's probably fine to not call that out explicitly.
